jax/interpreters/masking.py

```diff
@@ -104,30 +106,98 @@ class ShapeExpr(tuple):  # type ShapeExpr = [Poly]
   def __str__(self):
     return 'ShapeExpr({})'.format(', '.join(map(str, self)))
   def __getitem__(self, idx):
-    if type(idx) is int:
-      return super(ShapeExpr, self).__getitem__(idx)
-    else:
-      return ShapeExpr(super(ShapeExpr, self).__getitem__(idx))
+    poly = super(ShapeExpr, self).__getitem__(idx)
+    return poly if type(idx) is int else ShapeExpr(poly)
+
+
+def canonicalize_poly(p):
+  if onp.isscalar(p) and onp.issubdtype(onp.int, onp.array(p).dtype):
+    return constant_poly(p)
+
+  assert isinstance(p, Poly)
+  return p
+
+def canonicalize_polymorphic_shape(shape):
+  return ShapeExpr(map(canonicalize_poly, shape))
+
+def poly_without_zeros(d):
+  d = {mon: count for mon, count in d.items() if count != 0}
+
+  return constant_poly(0) if len(d) == 0 else Poly(d)
 
 class Poly(Counter):  # type Poly = Map Mon Int -- monomials to coeffs
-  def __mul__(p1, p2):
-    new_poly = Poly()
-    for (mon1, coeff1), (mon2, coeff2) in it.product(p1.items(), p2.items()):
+  def __add__(self, other):
+    d = self.copy()
+
+    for mon, count in canonicalize_poly(other).items():
+      d[mon] = d.get(mon, 0) + count
+
+    return poly_without_zeros(d)
```

Comment:
Can you say a bit about why you added this method (rather than relying on the one inherited from `Counter`)? (I assume it's to be able to call `poly_without_zeros`.)

Reply:
Tried this, but `Counter.__add__` doesn't work correctly with negative values. These are needed to express polynomials like `n-1`.

Reply:
That makes perfect sense, thanks for figuring it out and explaining!